packages/contracts/test/registry/ens-subdomain-registry.ts

```diff
@@ -0,0 +1,253 @@
+    } else {
+      fullDomain = subdomain + '.' + domain;
+    }
+
+    let tx = await ens
+      .connect(domainOwner)
+      .setSubnodeRecord(
+        namehash(domain),
+        labelhash(subdomain),
+        subdomainOwnerAddress,
+        resolver.address,
+        0
+      );
+    tx.wait();
+
+    // Verify that the subdomain is owned by the correct address
+    expect(await ens.owner(namehash(fullDomain))).to.equal(
+      subdomainOwnerAddress
+    );
+    // Verify that that the subdomain's resolver address is set correctly
+    expect(await ens.resolver(namehash(fullDomain))).to.equal(resolver.address);
+  }
+
+  beforeEach(async () => {
+    signers = await ethers.getSigners();
+
+    // Setup ENS with signers[0] owning the the ENS root node ('') and the resolver node ('resolver')
+    [ens, resolver] = await setupENS(signers[0]);
+
+    // create a DAO
+    const DAO = await ethers.getContractFactory('DAO');
+    dao = await DAO.deploy();
+    await dao.initialize(
+      '0x',
+      await signers[0].getAddress(),
+      ethers.constants.AddressZero
+    );
+
+    // Create the registrar to be tested
+    const ENSSubdomainRegistrar = await ethers.getContractFactory(
+      'ENSSubdomainRegistrar'
+    );
+    subdomainRegistrar = await ENSSubdomainRegistrar.deploy();
+  });
+
+  describe('initialize:', () => {
+    it('initializes if the node owner has approved the subdomain registrar address', async () => {
+      // Register the parent domain 'test' through signers[0] who owns the ENS root node ('') and make the signers[0] the owner
+      registerSubdomainHelper('test', '', signers[0], signers[0].address);
+
+      // Approve the subdomain registrar contract address to operate for signers[0] (who owns 'test')
+      await ens
+        .connect(signers[0])
+        .setApprovalForAll(subdomainRegistrar.address, true);
+
+      expect(
+        await subdomainRegistrar
+          .connect(signers[0])
+          .initialize(dao.address, ens.address, namehash('test'))
+      );
+    });
+
+    it('initializes if the subdomain registrar owns the node', async () => {
+      // Register the parent domain 'test' through signers[0] who owns the ENS root node ('') and make the subdomain registrar the owner
+      registerSubdomainHelper(
+        'test',
+        '',
+        signers[0],
+        subdomainRegistrar.address
+      );
+
+      expect(
+        await subdomainRegistrar
+          .connect(signers[0])
+          .initialize(dao.address, ens.address, namehash('test'))
+      );
+    });
+
+    it('does not initialize if the registrar is neither the domain node owner nor an approved operator of the domain node owner', async () => {
+      await expect(
+        subdomainRegistrar
+          .connect(signers[0])
+          .initialize(dao.address, ens.address, namehash('test'))
+      ).to.be.revertedWith(
+        customError(
+          'RegistrarUnauthorized',
+          ethers.constants.AddressZero,
```

Comment:
Since the conditional contains the `&&` operator (both conditions must be met)
https://github.com/aragon/core/blob/183ccaa5df7ec3585ffdc73dd65dd7ddd58a2ffd/packages/contracts/contracts/registry/ens/ENSSubdomainRegistrar.sol#L36-L38
it is not possible to separate them.

> Also, shouldn't we be trying to connect using another random signer?
`signers[0]` owns the ENS registries and if it fails for him, this is already sufficient. However, I agree we can also check this for another signer who does not own the registries

Reply:
I know, but I mean about the test, not about the contract code.

Instead of doing one assertion to test two things at the same time, I would always have two assertions, making each one of the if conditions trigger the `revert`. Otherwise we can't really tell that the revert is because of one or the other.

Reply:
We were testing ownership / authorization separately already and there is only one condition that can result in revert! Maybe there is a misunderstanding about the logic.

We currently have three separate tests: 
1. The registrar owns the parent domain (pass)
2. The registrar is approved by the owner of the parent domain (pass)
3. Neither 1. or 2. is true (revert)

The missing 4th case would be if the registrar is the owner of the parent domain AND approved itself (pass), 
which does not make sense to test (as we would need to hardcode this into the registrar contract so that the registrar is the msg.sender).

However, I see that the two negations make this less readable. I will change this to the equivalent expression
```
if ( !(nodeOwner == address(this) || _ens.isApprovedForAll(nodeOwner, address(this)) ) { 
  revert ...
} 
```

Reply:
As you preferred the flat version in your latest review, I reverted the commit.

Reply:
> Deny everything unless all checks are clean

> 1. If not "registrar owns the parent domain" => Revert
> 2. If not "registrar is approved by the owner of the parent domain" => Revert
> 3. Otherwise (by elimination) => pass

Maybe I am completely missing something, but what you describe here is a logical AND. However, we need a logical OR here.

If we revert
> "registrar owns the parent domain" => Revert`

because the owner does not own the parent domain, we won't be able to check if he/she is approved afterwards.

I did it the same way [as it is done in the original `ENSRegistry` contract and intentionally mimicked their `authorised` modifier](
https://github.com/ensdomains/ens-contracts/blob/083d29a2c50cd0a8307386abf8fadc217b256256/contracts/registry/ENSRegistry.sol#L19-L24):
```js
    modifier authorised(bytes32 node) {
        address owner = records[node].owner;
        require(owner == msg.sender || operators[owner][msg.sender]);
        _;
    }
```

Reply:
Oh well, my comment was more in terms of thinking about access control, not about this exact logical clause. I totally agree that this is an AND condition 💯 

But from a testing perspective, we should be checking the two conditions in isolation and not in combination.